Dotty with explicit nulls (and flow typing)

abeln opened the pull request:
This PR adds a `-Yexplicit-nulls` experimental flag which, if enabled, modifies the type system so that reference types are non-nullable. Nullability needs then to be expressed explicitly via unions (e.g. `String|Null`).

At a high level, the changes to the compiler are:
  - new type hierarchy for Null, as described
  - "translation layer" from Java types to Scala types, which tries to balance soundness and usability

abeln commented:
@odersky this is a version of #6344 without the flow typing. PTAL.

odersky commented:
I'll get to this ASAP

odersky commented on compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala:
```diff
@@ -0,0 +1,144 @@
+import dotty.tools.dotc.core.Types.{AndType, ClassInfo, ConstantType, OrType, Type, TypeBounds, TypeMap, TypeProxy}
+
+/** Defines operations on nullable types. */
+object NullOpsDecorator {
```
We should integrate `Types.OrNull` here and remove any duplications. Right now we have stripNull twice. I believe we can drop the one in `Types.OrNull`.

noti0na1 replied:
The `stripNull` in `Types.OrNull` is removed.

odersky commented on compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala:
```diff
@@ -0,0 +1,144 @@
+          if (rrhs.isNullType) llhs
+          else if (llhs.isNullType) rrhs
+          else tp.derivedOrType(llhs, rrhs)
+        case _ =>
```
Consider normalizing also in intersections, the way it is done in `OrNull.stripNull`.

noti0na1 replied:
Normalizing intersections is added.

odersky commented on compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala:
```diff
@@ -0,0 +1,144 @@
+            case _ => mapOver(tp)
+          }
+      }
+      val rem = RemoveNulls(self.widenDealias)
```
You can avoid the `diff` side effect be changing to:
```
  val self1 = self.widenDealias
  val rem = RemoveNulls(self1)
  if rem ne self1 then rem else self
```

noti0na1 replied:
The variable is removed

odersky commented on compiler/src/dotty/tools/dotc/core/Flags.scala:
```diff
@@ -450,8 +450,12 @@ object Flags {
+    // detect it in JavaNullInterop. However, JavaEnumValue is not initialized at this
+    // point, so we just make sure that all the "primitive" flags contained in JavaEnumValue
+    // are mentioned here as well.
+    Enum, StableRealizable)
```
How do we know that these are indeed never changed after loads? Assuming a flag as an after load flag that can nevertheless still be changed later produces a kind of race that we want to avoid. This will be checked once #7427 is in.

noti0na1 replied:
Removed.

odersky commented on compiler/src/dotty/tools/dotc/core/Symbols.scala:
```diff
@@ -383,6 +383,15 @@ trait Symbols { this: Context =>
+  /** Get a List of ClassSymbols which are either defined in current compilation
+   *  run or present on classpath.
+   */
+  def getClassesIfDefined(pathes: List[PreName]): List[ClassSymbol] =
```
Should pathes be paths?

noti0na1 replied:
Typo fixed

odersky commented on compiler/src/dotty/tools/dotc/core/TypeComparer.scala:
```diff
@@ -1606,9 +1607,18 @@ class TypeComparer(initctx: Context) extends ConstraintHandling[AbsentContext] w
+            // the Java argument has type 'Any', and the Scala argument has type 'Object' or
+            // 'Object|Null', depending on whether explicit nulls are enabled.
+            lazy val formal2IsObject =
+              if (ctx.explicitNulls) formal2.isNullableUnion && formal2.stripNull(ctx).isRef(ObjectClass)
```
It's better to make this (and the next lazy val a `def`. Local lazy vals allocate closures and are therefore expensive even if they are not used.

Also, I did not see what is gained here by the `isNullableUnion` test?

noti0na1 replied:
Changed to `def`; also use `OrNull` extractor to avoid normalizing twice.

odersky commented on compiler/src/dotty/tools/dotc/core/Types.scala:
```diff
@@ -598,11 +588,20 @@ object Types {
-          // supertype of `pre`.
-          go(tp.join)
+          if (ctx.explicitNulls && tp.isJavaNullableUnion) {
+            // Selecting `name` from a type `T|JavaNull` is like selecting `name` from `T`.
```
It seems isJavaNullable union and stripJavaNull do the same core operation twice. It might be more efficient to use a `OrJavaNull` extractor instead (which should succeed only under ctx.explicitNulls):
```
  tp match
    case OrJavaNull(tp1) => tp1.findMember(name, pre.stripJavaNull, required, excluded)
    case _ => go(tp.join)
```

noti0na1 replied:
`OrJavaNull` is added.

odersky commented on compiler/src/dotty/tools/dotc/core/Types.scala:
```diff
@@ -1066,21 +1065,60 @@ object Types {
+            case _ => orig.derivedOrType(tpe, defn.NullType)
+          }
+
+          // Test for nullable union that assumes the type has already been normalized.
```
This looks like the right version. There will be merge errors with add-flow-analysis, so this should replace the other version.

noti0na1 replied:
`widenUnion` is rewritten using `OrNull` .

noti0na1 replied:
This code is now deleted.

odersky commented on compiler/src/dotty/tools/dotc/transform/LazyVals.scala:
```diff
@@ -16,7 +16,7 @@ import dotty.tools.dotc.core.Types._
 import dotty.tools.dotc.transform.MegaPhase.MiniPhase
 import dotty.tools.dotc.transform.SymUtils._
-
+import dotty.tools.dotc.core.NullOpsDecorator._
```
Does not seem to be needed

noti0na1 replied:
Removed

odersky commented on compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala:
```diff
@@ -17,6 +17,7 @@ import config.Printers.patmatch
 import reporting.diagnostic.messages._
 import dotty.tools.dotc.ast._
 import util.Property._
+import NullOpsDecorator._
```
Does not seem to be needed

noti0na1 replied:
Removed

odersky commented on compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala:
```diff
@@ -14,6 +14,7 @@ import util.Spans._
 import reporting.diagnostic.messages.TypeTestAlwaysSucceeds
 import reporting.trace
 import config.Printers.{ transforms => debug }
+import NullOpsDecorator._
```
Does not seem to be needed

noti0na1 replied:
Removed

odersky commented on compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:
```diff
@@ -22,6 +22,7 @@ import reporting.diagnostic.messages._
 import reporting.trace
 import config.Printers.{exhaustivity => debug}
 import util.SourcePosition
+import NullOpsDecorator._
```
It would be good if Fengyun could review this part.

abeln replied:
Fengyun ok'ed the `Space` changes.

odersky commented:
@liufengyun Could you do a review of the changes to Space.scala? Thanks!

odersky commented on docs/docs/reference/other-new-features/explicit-nulls.md:
```diff
@@ -0,0 +1,296 @@
+title: "Explicit Nulls"
+---
+
+This proposal describes a modification to the Scala type system that makes reference types
```
If it's merged it's no longer a proposal, so this needs another introduction.

odersky replied:
Generally, this should be slightly revised so that it'a doc page for a Scala developer. This means providing more examples and simple explanations. There should be another page with more details that is linked to at the end of the doc page. I would defer all soundness discussions to this second page. 

So the first page should be and example driven use-level description, and the second page can then fill in spec, meta-properties such as soundness, and discussion of design tradeoffs.

The docs should also cover flow analysis and how flow analysis links up with typing. See my comment https://github.com/lampepfl/dotty/pull/6344#issuecomment-554062532.

odersky commented on docs/docs/reference/other-new-features/explicit-nulls.md:
```diff
@@ -0,0 +1,296 @@
+
+To make working with nullable values easier, we propose adding a few utilities to the standard library.
+So far, we have found the following useful:
+  - An extension method `.nn` to "cast away" nullability
```
Should use new extension method syntax

noti0na1 replied:
The definition is updated

sjrd commented on docs/docs/reference/other-new-features/explicit-nulls.md:
```diff
@@ -0,0 +1,296 @@
+
+Because of the unsoundness, we need to allow comparisons of the form `x == null` or `x != null`
+even when `x` has a non-nullable reference type (but not a value type). This is so we have an
+"escape hatch" for when we know `x` is nullable even when the type says it shouldn't be.
```
I am still convinced that this is a bad idea. We shouldn't expose an escape hatch to test for null under unsoundness. There is exactly 0 use case for this. If you observe an unsound `null`, you are already in the realm of buggy programs. There is no valid program that will need this.

And even if I'm wrong and someone really really needs an escape hatch for this, there is *anyway* another workaround, which is `(x: Any) == null`.

abeln replied:
We will revert this. The original motivation was that it would ease migration to explicit nulls, but I don't have the data to back up that assertion at the moment. We can always add things back if and when they're needed down the line.

abeln replied:
@odersky @sjrd where in Dotty can we find the logic that determines whether a comparison `a == b` is valid? For some reason, the following typechecks
```scala
val s: String = ???
if (s == null) {}
```
even though `s` has its `Eql` instance defined `Eql[String, String]`, but we haven't defined one for `Null`. All the logic we could find was in `Implicits.scala`, in `synthesizedEq`, but that seems to be related to finding the "fallback" instance `Eq[Any, Any]`.

Where should we restrict comparisons like the one above. Thanks.

noti0na1 replied:
I have found the logic in `Implicits.scala`, and I disallow the comparison between Object and Null for explicit nulls.

sjrd commented on docs/docs/reference/other-new-features/explicit-nulls.md:
````diff
@@ -0,0 +1,296 @@
+However, we also need to allow reference equality comparisons:
+```scala
+val x: String = null
+x eq null // ok: could return `true` because of unsoundness
````
See above, this shouldn't be allowed either.

noti0na1 replied:
This "special" `eq` has been removed.

sjrd commented on docs/docs/reference/other-new-features/explicit-nulls.md:
````diff
@@ -0,0 +1,296 @@
+We support this case via extension methods defined in the Predef:
+```scala
+def (x: AnyRef|Null) eq(y: AnyRef|Null): Boolean =
+  (x == null && y == null) || (x != null && x.eq(y))
````
Without flow typing, the last call to `x.eq(y)` will call back into this extension method, causing an infinite recursion, won't it? Even *with* flow typing, I would find this fragile. You should explicitly cast both `x` and `y` too `AnyRef` in that last comparison.

And since you need the casts anyway, you can take advantage of the type erasure to implement the whole method as just
```scala
... = x.asInstanceOf[AnyRef] eq y.asInstanceOf[AnyRef]
```

sjrd commented on library/src/dotty/DottyPredef.scala:
```diff
@@ -38,4 +38,32 @@ object DottyPredef {
+   *
+   *  Note that `.nn` performs a checked cast, so if invoked on a null value it'll throw an NPE.
+   */
+  def[T] (x: T|Null) nn: T =
```
Consider using `x.type & T` as result type for this extension method.

noti0na1 replied:
Changed to `x.type & T`

sjrd commented on library/src/dotty/DottyPredef.scala:
```diff
@@ -38,4 +38,32 @@ object DottyPredef {
+   *  `eq` method in `AnyRef`.
+   */
+  def (x: AnyRef|Null) eq(y: AnyRef|Null): Boolean =
+    (x == null && y == null) || (x != null && x.eq(y))
```
As mentioned above, this can be more efficient as:
```suggestion
    x.asInstanceOf[AnyRef] eq y.asInstanceOf[AnyRef]
```

abeln replied:
This method no longer exists.

sjrd commented on docs/docs/reference/other-new-features/explicit-nulls.md:
````diff
@@ -0,0 +1,296 @@
+```
+There are two options:
+  - `def foo(): String`
+  - `def foo(): String|Null`
````
There is also `String | JavaNull`. Have you considered that?

noti0na1 replied:
Section removed

odersky commented:
Since we will no longer pursue `NotNull` for now, I believe it's better to make `Null` a value class. But this could be done in a separate issue.

So, things to do: 

 - Merge #7556
 - Implement #7546 on top of it, using the `notNull` assertion outlined in `notNull.scala`, but as an `inline` function:
```
inline def $notNull[A](x: A | Null): x.type & A =
    x.asInstanceOf
```
One needs to verify that `$notNull` calls are indeed erased completely. I propose to put the method in the scala.compiletime package object.

abeln commented:
We're working on addressing the review comments. We'll implement the insertion of `notNull` in a separate PR for readability (but we'll wait until both PRs are accepted before committing either of them).

lrytz commented on docs/docs/reference/other-new-features/explicit-nulls.md:
```diff
@@ -0,0 +1,296 @@
+## Binary Compatibility
+
+Our strategy for binary compatibility with Scala binaries that predate explicit nulls
+is to leave the types unchanged and be compatible but unsound.
```
If I understand correclty, this section applies not only to Scala libraries compiled before explicit nulls were available, but also to new libraries compiled without `-Yexplicit-nulls` - which will probably be the majority for a while. It would be good to mention this explicitly.

abeln replied:
as suggestion above + add example

abeln commented:
@odersky can you clarify your comment in https://github.com/lampepfl/dotty/pull/7546#issuecomment-554346321 that `notNull` should be `inline`, given that we will remove it later during/after erasure?

abeln commented:
> @odersky can you clarify your comment in [#7546 (comment)](https://github.com/lampepfl/dotty/pull/7546#issuecomment-554346321) that `notNull` should be `inline`, given that we will remove it later during/after erasure? Thanks

@odersky friendly ping on the question above

odersky commented:
> @odersky can you clarify your comment in #7546 (comment) that notNull should be inline, given that we will remove it later during/after erasure? Thanks

My comment was that simply the fact of making it inline will eliminate it at erasure. No special logic should be needed!

noti0na1 commented:
@odersky I have a problem about the notNullInfo of `tpd.Typed` tree.

```scala
if (x != null) {
  ???
}
```

When typing the If tree above, the notNullInfo from the condition expression can be correctly passed to the body. 

```scala
if ((x != null): Boolean) {
  ???
}
```

However, if the condition expression is type-annotated like the example above, then the NotNullInfo is not passed. 

I checked the `expr1` variable of `ascription` function in `typedTyped`, it seems its NotNullInfo is empty.

The Apply tree `x.!=(null)` in the first example can generate correct NotNullInfo, but I don't understant why the Apply tree in the second example contains empty NotNullInfo. Thanks.

noti0na1 commented:
Hi @sjrd and @odersky ,
I have some problems about the design of `$notNull` function.

I think the overall ideas are:
1. Be able to track the TermRef according to the type
2. `$notNull` can be safely removed after Erasure.

The original definition is:

```scala
inline def $notNull[A](x: A | Null): x.type & A = x.asInstanceOf 
```

`x.type` allows us to check the TermRef to x, so we can track longer paths like `x.a.b`.

The first problem is: this type `x.type & A` doesn't denote a stable reference anymore. So I add `case tp: AndType => tp.tp1.isStable || tp.tp2.isStable` in `isStable` in `Types.scala`. I think it is ok to add this case, because the path is still idempotent. Am I correct?

The second problem is: this definition doesn't work well with `VarDef`s. When calling this function, the argument will be first assigned to a `ValDef`. The return type (`(T|Null)($1) & T`) will contain a unknown value in it which is not helpfull for the compiler and users.

Hence, I have to change the definition to this so I can specify the return type:

```scala
inline def $notNull[A, B](x: => A | Null): B = x.asInstanceOf
```

For any non-nullable val member or path, we can create a call like this in compiler: `$notNull[A, x.type & A](x)`. Since it is inlined, I think it is easier to not create the `$notNull` definition, just use `x.asInstanceOf[x.type & A]` directly.

Do you have any comments on these changes? Thanks.

sjrd commented:
I'm not sure I completely understand the problems you are facing, but the solutions look reasonable to me.

odersky commented:
> The first problem is: this type x.type & A doesn't denote a stable reference anymore. So I add case tp: AndType => tp.tp1.isStable || tp.tp2.isStable in isStable in Types.scala. I think it is ok to add this case, because the path is still idempotent. Am I correct?

Yes, that looks correct.

> For any non-nullable val member or path, we can create a call like this in compiler: $notNull[A, x.type & A](x). Since it is inlined, I think it is easier to not create the $notNull definition, just use x.asInstanceOf[x.type & A] directly.

I agree. Or, rather, use `defn.Any_typeCast`, which is rendered `x.$asInstanceOf$[x.type & A]`.
The difference between the two is that `Any_typeCast` is assumed to never fail, so an expression
using the cast can still be considered pure. We might even be able to declare it a path, which would
solve our expressiveness issue for paths. But that can be left until later.

So far, `Any_TypeCast` is only inserted by the pattern matcher, to cast a value after it was verified that it is an instance of the target type. But we could use it equally well for null typing, catching two birds with one stone.

smarter commented:
>> The first problem is: this type x.type & A doesn't denote a stable reference anymore. So I add case tp: AndType => tp.tp1.isStable || tp.tp2.isStable in isStable in Types.scala. I think it is ok to add this case, because the path is still idempotent. Am I correct?

> Yes, that looks correct.

That looks suspicious to me, it means `x.type & Nothing` or `x.type & { type T = <bad bounds>` would be stable and therefore usable as a path prefix, which would break soundness, wouldn't it ?

odersky commented:
> That looks suspicious to me, it means x.type & Nothing or x.type & { type T = <bad bounds> would be stable and therefore usable as a path prefix, which would break soundness, wouldn't it ?

Good point. I think we have to demand that `A` is known to have good bounds for this.

noti0na1 commented:
Hi @sjrd and @odersky ,

Since the change to implement flow-typing is not very large, we decided to include that part in this PR as well. We will update the title and description later.

I still have a problem with the `isTracked` in `Nullables.scala` and I am blocked by this.

For the mutable variable track, I think:
* If all the assignments of variable are at the same block as the definition, it is safe to track and use the `NotNullInfo`s in this block. 
* If there exist assignments in a closure, then it is not safe to track the variable anywhere. 
* Even if a closure doesn't contain assignments, it is still not safe to use `NotNullInfo`s in the closure, as the closure can be executed at the same time as the block containing the definition of the variable.

Originally, `Nullables.scala` uses the following code the check whether a` TermRef` is captured by a method.

```scala
def isTracked(ref: TermRef)(given Context) =
  ref.isStable
  || { val sym = ref.symbol
    sym.is(Mutable)
    && sym.owner.isTerm
    && sym.owner.enclosingMethod == curCtx.owner.enclosingMethod
    && sym.span.exists
    && curCtx.compilationUnit != null // could be null under -Ytest-pickler
    && curCtx.compilationUnit.assignmentSpans.contains(sym.span.start)
  }
```

This code works for the following example:

```scala
var x: String|Null = ???
def y = {
  if (x != null) {
    val a: String = x // error: x is captured by the closure of y, no longer tackable in the closure
  }
  x
}
if (x != null) {
  val a: String = x // ok  
}
```

But it doesn't work if we have `lazy val`:

```scala
lazy val y = {
  if (x != null) {
    val a: String = x // error: x is captured by the closure of y, no longer tackable in the closure
  }
  x
}
if (x != null) {
  val a: String = x // ok  
}
```

So I changed the line `sym.owner.enclosingMethod == curCtx.owner.enclosingMethod` in `isTracked` to:

```scala
if sym.owner != curCtx.owner then
  !curCtx.owner.is(Flags.Lazy) // not at the rhs of lazy ValDef
  && sym.owner.enclosingMethod == curCtx.owner.enclosingMethod
else true
```

But it still doesn't work for the by-name paramters:

```scala
abstract class F {
  def get(): String | Null
}

def f(x: => String | Null): F = new F {
  def get() = x
}

locally {
  var x: String|Null = ???
  val y: F = f {
    if (x != null) {
      val a: String = x // error: x is captured by the by-name paramter, which is transfored to closure later
    }
    x
  }
}
```

I can't find a good way to detect this case. Is there a general method to detect whether a `TermRef` is captured by some closure?

Thanks

abeln commented on docs/docs/internals/explicit-nulls.md:
```diff
@@ -0,0 +1,100 @@
+
+Java Interop
+------------
+TODO(abeln): add support for recognizing nullability annotations a la
```
Need to update.

abeln commented on docs/docs/internals/explicit-nulls.md:
```diff
@@ -0,0 +1,100 @@
+explicit nulls world.
+
+In order to nullify a member, we first pass it through a "whitelist" of symbols that need
+special handling (e.g. `constructors`, which never return `null`). If none of the "policies" in the
```
Remove language about "policies".

abeln commented on docs/docs/internals/explicit-nulls.md:
```diff
@@ -0,0 +1,100 @@
+
+In order to nullify a member, we first pass it through a "whitelist" of symbols that need
+special handling (e.g. `constructors`, which never return `null`). If none of the "policies" in the
+whitelist apply, we then process the symbol with a `TypeMap` that implements the following nullification
```
that implements the following nullification -> that implements the nullification function defined in XYZ

abeln replied:
remove function def below

abeln commented on docs/docs/internals/explicit-nulls.md:
```diff
@@ -0,0 +1,100 @@
+
+Working with Nullable Unions
+----------------------------
+Within `Types.scala`, we defined a few utility methods to work with nullable unions. All of these
```
update location

abeln replied:
mention extractors and change each description to single sentence without many details (otherwise, description will go out-of-date)

abeln commented on docs/docs/reference/other-new-features/explicit-nulls.md:
```diff
@@ -0,0 +1,296 @@
+title: "Explicit Nulls"
+---
+
+This proposal describes a modification to the Scala type system that makes reference types
```
proposal -> Explicit nulls is an opt-in feature that modifies the Scala type system...

abeln replied:
we need to talk about flow typing here

abeln commented on docs/docs/reference/other-new-features/explicit-nulls.md:
````diff
@@ -0,0 +1,296 @@
+1 == null       // error: Int is a value type
+```
+
+### Reference Equality
````
can be removed

abeln commented on docs/docs/reference/other-new-features/explicit-nulls.md:
```diff
@@ -0,0 +1,296 @@
+
+## Equality
+
+Because of the unsoundness, we need to allow comparisons of the form `x == null` or `x != null`
```
can only compare for equality nullable types
if want to compare non-null for some reason, can use cast (x: Any) != null

abeln replied:
also mention that can't use reference equality with null

abeln commented on docs/docs/reference/other-new-features/explicit-nulls.md:
````diff
@@ -0,0 +1,296 @@
+### Nullification Function
+
+We do the patching with a "nullification" function `n` on types:
+```scala
````
delete

abeln commented on docs/docs/reference/other-new-features/explicit-nulls.md:
```diff
@@ -0,0 +1,296 @@
+    `Box[T|JavaNull]|JavaNull`. This is needed because our nullability function is only applied (modularly) to the Java
+    classes, but not to the Scala ones, so we need a way to tell `Box` that it contains a nullable value.
+
+  * Rules 6, 7, and 8 just recurse structurally on the components of the type.
```
maybe delete

abeln commented on docs/docs/reference/other-new-features/explicit-nulls.md:
````diff
@@ -0,0 +1,296 @@
+  def foo(): String = ???
+}
+```
+There are two options:
````
remove "two options"

abeln commented on docs/docs/internals/explicit-nulls.md:
```diff
@@ -0,0 +1,100 @@
+---
```
Let's also describe flow typing here (how it's implemented)

abeln commented:
@odersky this is now ready for review (but see @noti0na1 's comment https://github.com/lampepfl/dotty/pull/7546#issuecomment-561354385)

odersky commented:
Hi @noti0na1,

I believe `isTracked` should be used only for determining whether a variable can be tracked at all. You need a separate predicate to check whether a reference to a trackable variable is used "in order" (in which case nullability flow info influences the type) or "out of order" (in which case we just use the declared type of the variable). Let's define a predicate `usedOutOfOrder` for the second case.

So, `usedOutOfOrder(x)` is `true` if the use appears

 1. under a lambda, or
 2. in a nested class, or
 3. in a nested method, or
 4. at the right-hand side of a lazy val, or
 5. as an argument of a by-name parameter

There's currently no implemented predicate to do that. We can implement (1) - (4) by looking at the context's owner chain. Let `x` be the referred to variable, and `xOwner` its owner. Then the use is out of order if some owner between `ctx.owner` (included) and `xOwner` (excluded) is a lazy val or a class or a method (i.e. the only owners that do not signal out of order are strict vals and vars).

The trickiest bit is, as you suggest, handling call-by-name arguments. Here, we do not know by looking at the owner chain `x` is part of a by-name argument or not. I have to think about what one can do in this case.

odersky commented:
Fixing by-name arguments looks really hard. 

I first considered this scheme:

 1. typecheck function arguments in general first under the assumption that the all arguments are by-value, taking flow info into account
 2. at the point where the function part of an application has been determined, check whether there are arguments for by-name parameters that make use of variable flow info, as witnessed by `x$asInstanceOf$[T]$` terms where `x` is a local variable.
 3. re-type any such arguments without flow info, and proceed with the resulting re-typed arguments.

This should be sound, but might miss a possible typing and produce an error instead.
```scala
def f(x: => String, y: Int) = 1
def f(x: String | Null, y: Long) = 2
var x: String | Null = ???
if x != null then 
  f(x, 1)
```
Here, if we had typechecked `x` on the last line as `x: String | Null`, we would pick the second variant of `f` and succeed. But since we typechecked it first as `x: String`, overloading resolution picks the first variant of `f`, which has the call by name parameter. retype-checking `x` will give `String | Null`,
and a type error results.

In such a case, the compiler could maybe offer an explanation why this happened, and suggest a remedy. A remedy could be explicitly ascribing the types of all mutable nullable references.

The problem with this scheme is that (2) above is ill defined. To determine an overloaded variant we might already have constructed and solved a constraint over the variant's type parameters. It looks really messy to try to undo that, and I'd rather not try.

odersky commented:
Here's another possibility:

 - We type check an argument without or with flow info, depending on whether the function part of the application is known to take a call-by-name parameter. If we know nothing about the function part, we assume call-by-value.
 - At the end of type checking an application, if the argument is known to be by-name, remove all `x.$asInstanceOf[T]$` casts in the argument where `x` is a mutable variable, and run the _type assigner_ (not the type checker!) again on the result. If this succeeds and gives a type that is still compatible with the formal parameter type, we are done.
 - Otherwise, issue an error saying that the argument cannot be treated as call-by-name since it contains flow-assumptions about mutable variables. As a remedy, suggest to wrap the argument in
a `scala.compiletime.byName(...)` call. Here, `byName` is defined as follws:
```scala
inline def byName[T](x: => T): T
```
Wrapping an argument with `byName` means that we know statically that it is passed to a by-name parameter, so it will be typechecked without flow info.

odersky commented on compiler/src/dotty/tools/dotc/typer/Typer.scala:
```diff
@@ -1554,16 +1570,6 @@ class Typer extends Namer
     typed(annot, defn.AnnotationClass.typeRef)
 
   def typedValDef(vdef: untpd.ValDef, sym: Symbol)(implicit ctx: Context): Tree = {
-    sym.infoOrCompleter match
```
I propose to move the code that was removed here into a separate method,
```scala
def adaptCreationContext(vdef: ValDef): Unit = ....
```

odersky commented on compiler/src/dotty/tools/dotc/typer/Typer.scala:
```diff
@@ -2216,14 +2222,31 @@ class Typer extends Namer
               // Keep preceding not null facts in the current context only if `mdef`
               // cannot be executed out-of-sequence.
-              case _: ValDef if !mdef.mods.is(Lazy) && ctx.owner.isTerm =>
-                ctx // all preceding statements will have been executed in this case
```
This is too complicated and also too inefficient. We should try to avoid unnecessary allocations here, and we should try to keep the code in typedStats reasonably short, since both factors affect runtime to a significant degree.
I propose to leave the code as is and just add a call to `adaptCreationContext(vdef)` in the first case clause. I.e.:
```scala
val defCtx = mdef match
  case vdef: ValDef if !mdef.mods.is(Lazy) && ctx.owner.isTerm =>
    // Keep preceding not null facts in the current context only if `mdef`
    // cannot be executed out-of-sequence.
    adaptCreationContext(vdef)
    ctx
  case _ =>
    ctx.withNotNullInfos(initialNotNullInfos)

noti0na1 replied:
The reason we didn't check `ValDef` and `Lazy` in our code is because we want it works for general DefTree. As long as it is not forward referenced, it is ok to use the NotNullIInfos. We also don't need to worry about mutable variables used out of order here, since it checked by function `usedOutOfOrder` later.

For example:
```scala
val s: String|Null = ???
if (s != null) {
  def f = // it is safe to use the fact s!=null
}
```

```scala
val s: String|Null = ???
val s = f
if (s == null) throw Exception
def f = {
  // it is not safe to use the fact s!=null
  // when f is first used, we don't know the fact about s
}
```

odersky replied:
Ah, yes, good points. The we should use  `adaptCreationContext(mdef)` for all definitions.

noti0na1 replied:
`adaptCreationContext(mdef: DefTree): Boolean` is added

abeln commented on compiler/src/dotty/tools/dotc/typer/Nullables.scala:
```diff
@@ -104,13 +105,57 @@ object Nullables with
+   *    }
+   *    if (x != null) {
+   *      // y can be called here, which break the fact
+   *      val a: String = x // error: x is captured and mutated by the closure, not tackable
```
tackable -> trackable
(here and in a few places below)

odersky commented:
Do you think we can get this ready by Monday? In that case we could still merge as part of the next Dotty release. That would be nice, since we are supposed to be in feature freeze afterwards. But if it does not work out, we'll just carve out an exception for it.

odersky commented:
See #7727 for the by-name part. It's not pretty, but I don't know of anything better. For testing it, we should base it on top of this PR.

hepin1989 commented on compiler/src/dotty/tools/dotc/core/Definitions.scala:
```diff
@@ -520,12 +538,16 @@ class Definitions {
-      case List(pt) => (pt isRef StringClass)
+      case List(pt) =>
+        val pt1 = if (ctx.explicitNulls) pt.stripNull else pt
+        pt1 isRef StringClass
```
pt1.isRef(StringClass)

noti0na1 replied:
Done

hepin1989 commented on compiler/src/dotty/tools/dotc/core/Definitions.scala:
```diff
@@ -793,6 +815,25 @@ class Definitions {
+    "org.checkerframework.checker.nullness.compatqual.NonNullDecl" ::
+    "org.jetbrains.annotations.NotNull" ::
+    "lombok.NonNull" ::
+    "io.reactivex.annotations.NonNull" :: Nil map PreNamedString)
```
javax.validation.constraints.NotNull

hepin1989 replied:
reactor.util.annotation. NonNull

hepin1989 replied:
org.springframework.lang.NonNull

hepin1989 replied:
org.springframework.lang.NonNullApi
org.springframework.lang.NonNullFields

noti0na1 replied:
Thanks, I have add them to the list

hepin1989 commented on compiler/src/dotty/tools/dotc/core/Definitions.scala:
```diff
@@ -793,6 +815,25 @@ class Definitions {
+  // improve the precision of type nullification.
+  // We don't require that any of these annotations be present in the class path, but we want to
+  // create Symbols for the ones that are present, so they can be checked during nullification.
+  @tu lazy val NotNullAnnots: List[ClassSymbol] = ctx.getClassesIfDefined(
```
NotNullAnnotations

noti0na1 commented on compiler/src/dotty/tools/dotc/typer/Nullables.scala:
````diff
@@ -104,13 +105,57 @@ object Nullables with
+   *    }
+   *    ```
+   *
+   *  See more examples in `tests/explicit-nulls/neg/var-ref-in-closure.scala`.
````
Adding `usedOutOfOrder` is done

noti0na1 commented on compiler/src/dotty/tools/dotc/core/Definitions.scala:
```diff
@@ -520,12 +538,16 @@ class Definitions {
-      case List(pt) => (pt isRef StringClass)
+      case List(pt) =>
+        val pt1 = if (ctx.explicitNulls) pt.stripNull else pt
+        pt1 isRef StringClass
```
Done

noti0na1 commented on compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala:
```diff
@@ -0,0 +1,143 @@
+     *      the same type as `self`, but where all instances of `Null` (`JavaNull`) in the union
+     *      have been removed. Then the result is `self2 | Null` (`self2 | JavaNull`).
+     */
+     def normNullableUnion(implicit ctx: Context): Type = {
```
`normNullableUnion` is removed; I rewrite `stripNull`, and now `stripJavaNull` uses  `stripNull`

noti0na1 commented on compiler/src/dotty/tools/dotc/core/Symbols.scala:
```diff
@@ -383,6 +383,15 @@ trait Symbols { this: Context =>
+   *  run or present on classpath.
+   */
+  def getClassesIfDefined(paths: List[PreName]): List[ClassSymbol] =
+    paths.foldLeft(List.empty){ case (acc, path) => getClassIfDefined(path) match {
```
Done

noti0na1 commented on compiler/src/dotty/tools/dotc/core/Types.scala:
```diff
@@ -600,11 +605,20 @@ object Types {
-          // lots of places. The present strategy is instead of widen `tp` using `join` to be a
-          // supertype of `pre`.
-          go(tp.join)
+          tp match {
```
Moved to a separate function

noti0na1 commented on compiler/src/dotty/tools/dotc/core/Types.scala:
```diff
@@ -2915,23 +2939,41 @@ object Types {
-      val tp1 = stripNull(tp)
+    def unapply(tp: Type)(given ctx: Context): Option[Type] =
+    if (ctx.explicitNulls) {
+      val tp1 = tp.stripNull
```
Done, `normNullableUnion` is removed; mutable variables are not used anymore

noti0na1 commented on compiler/src/dotty/tools/dotc/typer/ConstFold.scala:
```diff
@@ -12,6 +12,7 @@ import Names._
 import StdNames._
 import Contexts._
 import Nullables.{CompareNull, TrackedRef}
+import NullOpsDecorator._
```
Removed

noti0na1 commented on compiler/src/dotty/tools/dotc/typer/Nullables.scala:
```diff
@@ -104,13 +105,57 @@ object Nullables with
          && sym.owner.isTerm
-         && sym.owner.enclosingMethod == curCtx.owner.enclosingMethod
+         && ( sym.owner == curCtx.owner
+          || !curCtx.owner.is(Flags.Lazy) // not at the rhs of lazy ValDef
```
`usedOutOfOrder` is added

noti0na1 commented on compiler/src/dotty/tools/dotc/typer/Nullables.scala:
```diff
@@ -254,7 +299,13 @@ object Nullables with
+          // If the type of rhs is nullable (`T|Null` or `Null`), then the nullability of the
+          // lhs variable is no longer trackable. We don't need to check whether the type `T`
+          // is correct here, as typer will check it.
+          tree.withNotNullInfo(NotNullInfo(Set(), Set(ref)))
```
Add `ref` to the LHS is done

noti0na1 commented on compiler/src/dotty/tools/dotc/typer/Typer.scala:
```diff
@@ -347,6 +348,16 @@ class Typer extends Namer
   }
 
+  // If `tree`'s type is a `TermRef` identified by flow typing to be non-null, then
+  // cast away `tree`s nullability. Otherwise, `tree` remains unchanged.
```
Add example

noti0na1 commented:
Hi @odersky , I just pushed my latest changes which solve the comments.

odersky commented on compiler/src/dotty/tools/dotc/typer/Typer.scala:
```diff
@@ -2279,6 +2288,34 @@ class Typer extends Namer
   }
 
+  /** Tries to adapt NotNullInfos from creation context to the DefTree,
+   *  returns whether the adaption is successed. The adaption only success if the
```
... returns whether the adaption took place. An adaption only takes place if the

noti0na1 replied:
Done

odersky commented:
Except for these minor details, looks good now!

noti0na1 commented on compiler/src/dotty/tools/dotc/typer/Nullables.scala:
```diff
@@ -161,6 +182,60 @@ object Nullables with
+      val refSym = ref.symbol
+      val refOwner = refSym.owner
+
+      @tailrec def usedWithinClosure(s: Symbol): Boolean =
```
Renamed

noti0na1 commented on compiler/src/dotty/tools/dotc/typer/Nullables.scala:
```diff
@@ -254,7 +329,15 @@ object Nullables with
+          // is correct here, as typer will check it.
+          tree.withNotNullInfo(NotNullInfo(Set(), Set(ref)))
+        else
+          // otherwise, we know the variable will have a non-null value
```
The case is add

noti0na1 commented on compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala:
```diff
@@ -0,0 +1,86 @@
+
+    /** Is self (after widening and dealiasing) a type of the form `T | Null`? */
+    def isNullableUnion(implicit ctx: Context): Boolean = {
+      val striped = self.stripNull()
```
typos are fixed

noti0na1 commented:
@odersky Thanks for all the comments. I have solved the comments.
